Fix flaky Perforce integration tests

mollylogue opened the pull request:
Okay, this change might seem a bit odd, but I promise there's a reason. (Also if there are any ideas on how to do this in an alternate/better way, I am extremely open to them). 
Context: The source of the flakiness is that there's a race condition where the Perforce authz provider hasn't yet been added to the [list of providers](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@7853ce8bb9dc0802b1629add4e29c9ec607d5768/-/blob/internal/database/repos_perm.go?L21:47-21:59), and so we bypass doing any extra auth checking. At first, I was directly calling `authz.GetProviders` in the test, to validate that the provider was there, but this was always empty. 
The list of authz providers is initialized and kept up to date in various different [locations in the code](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+authz.SetProviders+-file:_test%5C.go%24+&patternType=standard), and essentially it's the responsibility of whatever package is using this list to make sure the list is up to date (I don't love this approach, but it's the way things are currently) 
The place where it really matters for the purpose of graphql tests is [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@7853ce8bb9dc0802b1629add4e29c9ec607d5768/-/blob/enterprise/cmd/frontend/internal/authz/init.go?L191:10-191:22) where it's refreshed every 5 seconds. I basically just added a graphql api to get the list of authz providers that currently exist. I'm happy to re-think how we populate this list, keep it up to date, and generally handle this global state, but I went with this approach for now to un-block the tests. 

## Test plan
Ensure the integration tests pass without flakiness.

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

mollylogue commented on dev/gqltest/sub_repo_permissions_test.go:
```diff
@@ -337,6 +329,24 @@ func syncUserPerms(t *testing.T, userID, userName string) {
 		t.Fatal("Waiting for user permissions to be synced:", err)
 	}
+	// Wait up to 30 seconds for Perforce to be added as an authz provider
+	err = gqltestutil.Retry(30*time.Second, func() error {
```
In theory this should never be longer than 5 seconds.

pjlast replied:
Should this part not happen before the `err := client.ScheduleUserPermissionsSync(userID)` line? Or does it not matter here? Feels weird that we're only confirming the authz provider after the user permissions have been synced.

kopancek replied:
Should we instead use exponential backoff strategy? The current [`Retry` implementation](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/gqltestutil/helper.go?L32) retries every 100ms, which is not ideal. Not sure how it would impact the scope of this PR, but that might be a better strategy for retries like this.

mollylogue replied:
So, that permissions sync will be using a different list of authz providers (in theory the same, just initialized/kept up to date in a different place in the code). I haven't run into that causing any flakiness in the tests. 

Wrt the backoff strategy, I was just using the pattern from the other gql tests. I'm not sure if there are any problems or flakiness that would arise if we switched to exponential backoff, but it's maybe worth exploring at some point.

ryanslade commented:
Pulled in the IAM team too in case they have any ideas.

ryanslade commented on dev/gqltest/sub_repo_permissions_test.go:
```diff
@@ -337,6 +329,24 @@ func syncUserPerms(t *testing.T, userID, userName string) {
 	}
+	// Wait up to 30 seconds for Perforce to be added as an authz provider
+	err = gqltestutil.Retry(30*time.Second, func() error {
+		authzProviders, err := client.AuthzProviders()
```
Why can't we use the existing `authz.GetProviders` function instead? 

https://github.com/sourcegraph/sourcegraph/blob/ce0233f5b66eece4c79a976ddce8d2e6b7a4d9d3/internal/authz/register.go#L57

ryanslade replied:
I'm not following why we need to wrap it in GraphQL, is it because it's only set up correctly in the frontend and not in tests?

sashaostrikov reviewed (approved):
LGTM!

The only concern that I have is that we are exposing new gql method in order to fix the test and whether it should be made accessible only by site admins?

ryanslade commented:
My only thought is that maybe we can try and get config via `conf.Get()` and set the providers from config like the other code does?

unknwon commented on cmd/frontend/graphqlbackend/authz.graphql:
```diff
@@ -131,6 +131,11 @@ extend type Query {
+    """
+    Returns a list of the types of authz providers that have been configured and will be used for syncing permissions.
+    """
+    authzProviders: [String!]!
```
```suggestion
    authzProviderTypes: [String!]!
```

nit

mollylogue replied:
You're right about that. I should change the docstring to reflect that. It's the list of authz providers that will be used in any query coming from the frontend to determine if the user has access or not.

mollylogue replied:
For the purposes of this test, we only care about the list of authz providers used in the frontend. The flakiness was happening when the list wasn't yet updated there, and we were bypassing any authz checking entirely in that case (which happened about 1 in 10 times)

mollylogue replied:
This new gql api is basically the equivalent of the repo-updater debug endpoint that you posted, just for the frontend. The state of the authz providers list could potentially differ between the two for a few seconds, causing test flakiness if we try to use the endpoint from repo-updater.

unknwon replied:
If our test cases are targeting the read path (checking perms), totally on board with the approach in the PR. I couldn't really in detail anymore though, if this set of test setup expects the background perms syncing to do anything (if not, then good!).

unknwon replied:
Note: no intention to block the PR, purely jumping on for discussions 😛

mrnugget replied:
> This new gql api is basically the equivalent of the repo-updater debug endpoint that you posted, just for the frontend.

Would it make sense then to tag this query as "for internal use only", like [some of the other queries we have](https://sourcegraph.sourcegraph.com/search?q=file%3Agraphql%24+internal&groupBy=repo)?

kopancek replied:
What if there is more than one provider of the same type? What is returned in that case? E.g. if we have 2 github oauth providers configured, what is the output of this query? `['github']` or `['github', 'github']`?

This is a configuration that is used sometimes (even in our Cloud Managed Instances, but as far as I know we have no tests for that...

mollylogue replied:
Honestly this is only being used for this test, to check the list of authz providers that the frontend is using. I'll mark the query as internal as @mrnugget suggested to clarify that. If we need to provide more details about the providers in the list (i.e. the url, etc) we can add that if and when we need it.

unknwon reviewed (commented):
Just wanna call out based on my understanding that this new query would only return authz providers from whichever the `frontend` that is taking your request (checking perms), and is not technically equivalent to authz providers that are used for syncing perms (as the docstring suggests).

In case you missed it, there is a debug endpoint from repo-updater that lists authz providers that uses for syncing perms: https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph@2b3a2db549c1c93caca63daf54dcfa1f6873eca6/-/blob/cmd/repo-updater/shared/main.go?L243:52-243:77

mollylogue commented:
> My only thought is that maybe we can try and get config via `conf.Get()` and set the providers from config like the other code does?

@ryanslade If we added code in this package to set the providers, it would not resolve the race condition which is causing the tests to be flaky (since we need to know the state of the list of providers as initialized by the frontend package).

ryanslade commented:
> > My only thought is that maybe we can try and get config via `conf.Get()` and set the providers from config like the other code does?
> 
> @ryanslade If we added code in this package to set the providers, it would not resolve the race condition which is causing the tests to be flaky (since we need to know the state of the list of providers as initialized by the frontend package).

Is the frontend manually setting the providers to something different to what would get set by the providers fetched from config? When services other than frontend call `conf.Get()` they're actually making an http call out to frontend to fetch the config. 

All I'm trying to avoid here is the creation of the new GraphQL queries, if that's not possible then don't block the merge :)

mrnugget commented on dev/gqltest/sub_repo_permissions_test.go:
```diff
@@ -337,6 +329,24 @@ func syncUserPerms(t *testing.T, userID, userName string) {
+		return gqltestutil.ErrContinueRetry
+	})
+	if err != nil {
+		t.Fatal("Waiting for user permissions to be synced:", err)
```
Is that the correct error message? The comment says we're waiting for Perforce to be added as authz provider.

I know this sounds nitpicky, but these errors are crucial when debugging why a test fails or is flaky

mrnugget commented on dev/gqltest/sub_repo_permissions_test.go:
```diff
@@ -337,6 +329,24 @@ func syncUserPerms(t *testing.T, userID, userName string) {
+	err = gqltestutil.Retry(30*time.Second, func() error {
+		authzProviders, err := client.AuthzProviders()
+		if err != nil {
+			t.Fatal(err)
```
Is that correct here or should you `return gqltestutil.ErrContinueRetry`? Also, does this need a error message too? (`t.Fatal("failed to fetch authzProviders", err)`)

mollylogue replied:
I don't think we want to continue retrying if there's an error fetching the authz providers. But you're right, we should wrap the error for more context.

mrnugget commented on internal/gqltestutil/permissions.go:
```diff
@@ -162,6 +162,23 @@ query BitbucketProjectPermissionJobs($projectKeys: [String!], $status: String, $
 }
 
+func (c *Client) AuthzProviders() ([]string, error) {
+	const query = `
+query {
+	authzProviders
+}`
```
Nitpick, but I do think multi-line strings with funny intendation are uglier than a single line:

```suggestion
	const query = `query { authzProviders }`
```

kopancek commented on enterprise/cmd/frontend/internal/authz/resolvers/resolver.go:
```diff
@@ -457,6 +457,15 @@ func (r *Resolver) AuthorizedUsers(ctx context.Context, args *graphqlbackend.Rep
 	}, nil
 }
 
+func (r *Resolver) AuthzProviders() []string {
```
I would like to see a unit test for this resolver, as it's a new functionality added and by convention we usually test every resolver / exposed graphql entity.

mollylogue commented:
> > > My only thought is that maybe we can try and get config via `conf.Get()` and set the providers from config like the other code does?
> > 
> > 
> > @ryanslade If we added code in this package to set the providers, it would not resolve the race condition which is causing the tests to be flaky (since we need to know the state of the list of providers as initialized by the frontend package).
> 
> Is the frontend manually setting the providers to something different to what would get set by the providers fetched from config? When services other than frontend call `conf.Get()` they're actually making an http call out to frontend to fetch the config.
> 
> All I'm trying to avoid here is the creation of the new GraphQL queries, if that's not possible then don't block the merge :)

I think what you're referring to is the _other_ kind of auth providers, if I'm not mistaken. Not the authz providers. I tried fetching that list in the test and it was empty, if I remember correctly.

mollylogue commented:
@ryanslade I believe you're referring to [this list](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/schema/schema.go?L114-121) of auth providers, which unfortunately named the same, but serves a different purpose.

ryanslade commented:
> @ryanslade I believe you're referring to [this list](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/schema/schema.go?L114-121) of auth providers, which unfortunately named the same, but serves a different purpose.

I'm talking about the providers that come back from this function:
https://github.com/sourcegraph/sourcegraph/blob/ad2e480c235a818e1dc957ebc1089a8f99c3bbee/enterprise/internal/authz/authz.go#L38

This will load providers from config, which will go via the frontend. But, if these are not the providers you mean then please ignore me :) It could also be that they *are* the providers you want but that the values set in config are no longer valid in which case getting them from GraphQL makes sense.

unknwon commented:
> > @ryanslade I believe you're referring to [this list](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/schema/schema.go?L114-121) of auth providers, which unfortunately named the same, but serves a different purpose.
> 
> I'm talking about the providers that come back from this function:
> 
> https://github.com/sourcegraph/sourcegraph/blob/ad2e480c235a818e1dc957ebc1089a8f99c3bbee/enterprise/internal/authz/authz.go#L38
> 
> This will load providers from config, which will go via the frontend. But, if these are not the providers you mean then please ignore me :) It could also be that they _are_ the providers you want but that the values set in config are no longer valid in which case getting them from GraphQL makes sense.

Authz providers are loaded from `external_services` table not site config? i.e. not from `conf.Get`/frontend but directly back from DB.

ryanslade commented:
> > > @ryanslade I believe you're referring to [this list](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/schema/schema.go?L114-121) of auth providers, which unfortunately named the same, but serves a different purpose.
> > 
> > 
> > I'm talking about the providers that come back from this function:
> > https://github.com/sourcegraph/sourcegraph/blob/ad2e480c235a818e1dc957ebc1089a8f99c3bbee/enterprise/internal/authz/authz.go#L38
> > 
> > This will load providers from config, which will go via the frontend. But, if these are not the providers you mean then please ignore me :) It could also be that they _are_ the providers you want but that the values set in config are no longer valid in which case getting them from GraphQL makes sense.
> 
> Authz providers are loaded from `external_services` table not site config? i.e. not from `conf.Get`/frontend but directly back from DB.

Ok, I think the naming is confusing me then. `FromConfig` in this instance means "from external service config?"

unknwon commented:
> > > > @ryanslade I believe you're referring to [this list](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/schema/schema.go?L114-121) of auth providers, which unfortunately named the same, but serves a different purpose.
> > > 
> > > 
> > > I'm talking about the providers that come back from this function:
> > > https://github.com/sourcegraph/sourcegraph/blob/ad2e480c235a818e1dc957ebc1089a8f99c3bbee/enterprise/internal/authz/authz.go#L38
> > > 
> > > This will load providers from config, which will go via the frontend. But, if these are not the providers you mean then please ignore me :) It could also be that they _are_ the providers you want but that the values set in config are no longer valid in which case getting them from GraphQL makes sense.
> > 
> > 
> > Authz providers are loaded from `external_services` table not site config? i.e. not from `conf.Get`/frontend but directly back from DB.
> 
> Ok, I think the naming is confusing me then. `FromConfig` in this instance means "from external service config?"

Yes 😅

ryanslade commented:
@unknwon That's very confusing, especially when the doc comment mentions site config:

https://github.com/sourcegraph/sourcegraph/blob/main/enterprise/internal/authz/authz.go#L28

unknwon commented:
> @unknwon That's very confusing, especially when the doc comment mentions site config:
> 
> [https://github.com/sourcegraph/sourcegraph/blob/main/enterprise/internal/authz/authz.go#L28](https://github.com/sourcegraph/sourcegraph/blob/main/enterprise/internal/authz/authz.go?rgh-link-date=2022-11-03T12%3A02%3A52Z#L28)

The docstring isn't entirely wrong, because a reasonable setup for repo perms is to have both authN providers and authZ providers. The former is often referred as "auth provider" as shorthand and is configured in site config, the latter is configured in external services.

mollylogue commented:
Hey folks, I believe I've addressed the PR comments/concerns that y'all raised. Pls feel free to take another look (request more changes or approve) before I merge this.
